packages/cel/src/checker/scopes.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright 2024-2025 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { CelFunc } from "../func.js";
+import type { CelIdent } from "../ident.js";
+
+/**
+ * Group is a set of Decls that is pushed on or popped off a Scopes as a unit.
+ * Contains separate namespaces for identifier and function Decls.
+ */
+export class Group {
+  constructor(
+    public readonly idents: Map<string, CelIdent> = new Map(),
+    public readonly functions: Map<string, CelFunc> = new Map(),
+  ) {}
+
+  /**
+   * Creates a new Group instance with a shallow copy of the variables and functions.
+   */
+  copy(): Group {
+    return new Group(new Map(this.idents), new Map(this.functions));
+  }
+}
+
+/**
+ * Scopes represents nested Decl sets where the Scopes value contains a Groups containing all
+ * identifiers in scope and an optional parent representing outer scopes.
+ * Each Groups value is a mapping of names to Decls in the ident and function namespaces.
+ * Lookups are performed such that bindings in inner scopes shadow those in outer scopes.
+ */
+export class Scopes {
+  constructor(
+    public readonly scopes = new Group(),
+    public readonly parent?: Scopes,
+  ) {
+    this.scopes = scopes;
+    this.parent = parent;
+  }
```

Comment:
Can we call this a `Scope`, singular, and eliminate the `Group` abstraction altogether? Also, can we eliminate mutation from this class?